src/rosenfeld_fmt.cc

```diff
@@ -434,35 +431,69 @@ void RosenfeldFMT::computeSphericalDerivative(std::shared_ptr<State> state)
-                for (int ig_idx=lower_bin; ig_idx < split_bin; ++ig_idx)
+                
+                double n = 100;
+                double x;
+                double dr;
+                
+                //Trapezoidal integral 1
+                dr = split/n;
+                x = lower;
+                
+                for (int ig_idx=0; ig_idx <=n; ++ig_idx)
                     {
-                    const auto rig = mesh->center(ig_idx);
-                    dF_dn3(idx) += 4. * M_PI * dphi_dn3(ig_idx) * (rig * rig) * dr;
+                    const auto dphi_dn3_ig = mesh->interpolate(x, dphi_dn3);
+                    const double factor = (ig_idx == 0 || ig_idx == n) ? 0.5 : 1.0;
+                    dF_dn3_1 += factor * 4. * M_PI * dphi_dn3_ig * (x * x) ;
+                    x += dr;
                     }
-                for (int ig_idx=split_bin; ig_idx < upper_bin; ++ig_idx)
+                    
+                //Trapezoidal integral 2
+                dr = (upper-split)/n;
+                x = split;                
+                
+                for (int ig_idx=0; ig_idx <=n; ++ig_idx)
                     {
-                    const auto rig = mesh->center(ig_idx);
-                    const double kernel_w3 = rig * (M_PI/r) * (sq_R - (r-rig)*(r-rig));
-                    const double kernel_w2 = rig * (2.*M_PI*R/r);
+                    const double kernel_w3 = x * (M_PI/r) * (sq_R - (r-x)*(r-x));
+                    const double kernel_w2 = x * (2.*M_PI*R/r);
                     // minus sign for wv2 kernel to account for swap of variables in convolution with odd function
-                    const double kernel_wv2 = -rig * (M_PI/(r*r)) * (R*R + r*r - rig*rig);
+                    const double kernel_wv2 = -x * (M_PI/(r*r)) * (R*R + r*r - x*x);
                     double kernel_w0, kernel_w1, kernel_wv1;
                     computeProportionalByWeight(kernel_w0, kernel_w1, kernel_wv1, kernel_w2, kernel_wv2, R);
-
-                    dF_dn0(idx) += dphi_dn0(ig_idx)* kernel_w0 * dr;
-                    dF_dn1(idx) += dphi_dn1(ig_idx)* kernel_w1 * dr;
-                    dF_dn2(idx) += dphi_dn2(ig_idx)* kernel_w1 * dr;
-                    dF_dn3(idx) += dphi_dn3(ig_idx) * kernel_w3 * dr;
-                    dF_dnv1(idx) += dphi_dnv1(ig_idx) * kernel_wv1 * dr;
-                    dF_dnv2(idx) += dphi_dnv2(ig_idx) * kernel_wv2 * dr;
+                    
+                    const double factor = (ig_idx == 0 || ig_idx == n) ? 0.5 : 1.0;
+                    double dphi_dn0_ig = mesh->interpolate(x, dphi_dn0);
+                    double dphi_dn1_ig = mesh->interpolate(x, dphi_dn1);
+                    double dphi_dn2_ig = mesh->interpolate(x, dphi_dn2);
+                    double dphi_dn3_ig = mesh->interpolate(x, dphi_dn3);
+                    double dphi_dnv1_ig = mesh->interpolate(x, dphi_dnv1);
+                    double dphi_dnv2_ig = mesh->interpolate(x, dphi_dnv2);
```

Comment:
I would use `const auto` here, `const` since we know this will not change and `auto` since we want it to have the return type from `interpolate` (which will match the data type).
```suggestion
                    const auto dphi_dn0_ig = mesh->interpolate(x, dphi_dn0);
                    const auto dphi_dn1_ig = mesh->interpolate(x, dphi_dn1);
                    const auto dphi_dn2_ig = mesh->interpolate(x, dphi_dn2);
                    const auto dphi_dn3_ig = mesh->interpolate(x, dphi_dn3);
                    const auto dphi_dnv1_ig = mesh->interpolate(x, dphi_dnv1);
                    const auto dphi_dnv2_ig = mesh->interpolate(x, dphi_dnv2);
```